synapse/config/_util.py

```diff
@@ -86,7 +86,8 @@ def parse_and_validate_mapping(
     try:
         # type-ignore: mypy doesn't like constructing `Dict[str, model_type]` because
         # `model_type` is a runtime variable. Pydantic is fine with this.
-        instances = parse_obj_as(Dict[str, model_type], config)  # type: ignore[valid-type]
+        adapter = TypeAdapter(Dict[str, model_type])  # type: ignore[valid-type]
+        instances = adapter.validate_python(config, strict=True)
```

Comment:
I haven't looked very thoroughly at Pydantic 2.0, but IIRC this function is intended for validating JSON, and  I would expect us to use `validate_json` here rather than `validate_python`. (TODO: are there differences according to https://docs.pydantic.dev/2.0/usage/conversion_table/ ?)

Reply:
As I have not worked on this codebase before I can only follow my own assumptions here :sweat_smile: 

Is `config` at this point still bytes or a JSON string? I can certainly change it.

Reply:
FWIW, from running the tests it appears that it makes no difference here. Hower, since `JsonDict` already is a `Dict[str, Any]` I believe `validate_python` to be the more correct choice.